File: arrays/c#/dynamic_array.cs
public class DynamicArray<T>
{
    private long Capacity { get; set; }
    public long Length { get; private set; }

    private T[] Items;

    public DynamicArray()
    {
        Capacity = 1; Length = 0; Items = new T[1];
    }

    public DynamicArray(long capacity)
    {
        Capacity = capacity; Length = 0; Items = new T[capacity];
    }

    // Using in keywork to avoid coping large structs
    public void Add(in T item)
    {
        if (Length == Capacity)
        {
            Capacity *= 2;
            T[] extendedItems = new T[Capacity];
            Array.Copy(Items, extendedItems, Length);
            Items = extendedItems;
        }

        Items[Length] = item;
        Length++;
    }

    public void Remove()
    {
        if (Length == 0)
        {
            throw new InvalidOperationException("Array is Empty");
        }

        Items[Length - 1] = default!;
        Length--;   
    }

    public T this[long index]
    {
        get
        {
            if (index < 0 || index >= Length)
                throw new ArgumentOutOfRangeException($"Index {index} out of range");

            return Items[index];
        }

        set
        {
            if (index < 0 || index >= Length)
                throw new ArgumentOutOfRangeException($"Index {index} out of range");

            Items[index] = value;
        }
    }

    public DynamicArray<T> Copy()
    {
        DynamicArray<T> newArray = new(Capacity)
        {
            Length = Length,
        };

        Array.Copy(Items, newArray.Items, Length);
        return newArray;
    }

};

File: arrays/c+/dynamic_array.cpp
#include <iostream>
#include <stdexcept>
using namespace std;

template <typename T>
class DynamicArray
{
    private: 
        long Capacity;
        T* Items;

    public:
        long Length;

        DynamicArray() : Capacity(1), Length(0)
        {
            Items = new T[1];
        }

        DynamicArray(long capacity) : Capacity(capacity), Length(0)
        {
            Items = new T[capacity];
        }

        DynamicArray(const DynamicArray& arr) : Capacity(arr.Capacity), Length(arr.Length)
        {
            Items = new T[Capacity];
            for (long i = 0; i < Length; i++) Items[i] = arr.Items[i];
        }

        DynamicArray& operator=(const DynamicArray& arr)
        {
            if (this == &arr) return *this;

            delete[]Items;

            Capacity = arr.Capacity;
            Length = arr.Length;
            Items = new T[arr.Capacity];
            for (long i = 0; i < Length; i++) Items[i] = arr.Items[i];

            return *this;
        }



        void Add(const T& item)
        {
            if (Length == Capacity)
            {
                Capacity *= 2;
                T* extendedItems = new T[Capacity];

                for (long i = 0; i < Length; i++) extendedItems[i] = Items[i];

                delete[] Items;
                Items = extendedItems;
            }

            Items[Length] = item;
            Length++;
        }

        void Remove()
        {
            if (Length == 0)
            {
                throw out_of_range("Array is empty");
            }

            Items[Length - 1] = T{};
            Length--;
        }

        T& operator[](long index) 
        {
            if (index < 0 || index >= Length)
                throw out_of_range("Index out of range");

            return Items[index];
        }

        const T& operator[](long index)
        {
            if (index < 0 || index >= Length)
                throw out_of_range("Index out of range");

            return Items[index];
        }

      

        ~DynamicArray()
        {
            delete[] Items;
        }
};

